### PR TITLE
shape inference return status

### PR DESCRIPTION
**Pull request by linkerzhang**

### CODE REVIEW

File: onnx/defs/experiments/defs.cc [hunk 1]
```diff
@@ -175,25 +181,31 @@ ONNX_OPERATOR_SCHEMA(GivenTensorFill)
...
+		if (!hasInputShape(ctx, 0)) {
+			return Status::OK();
+		}
+
+		std::vector<int64_t> extra_shape;
+		getRepeatedAttribute(ctx, "extra_shape", extra_shape);
+		TensorShapeProto shape = ctx.getInputType(0)->tensor_type().shape();
+		for (auto extra_dim_val : extra_shape) {
+			if (extra_dim_val < 0) {
+				return Status(INFERENCE, INVALID_PROTOBUF, MakeString("extra_shape specified: ", extra_dim_val, " should not be negtive."));
```
Comment (gramalingam) [inline onnx/defs/experiments/defs.cc]: When should we use "INVALID_PROTOBUF" and when "INVALID_ARGUMENT"?

File: onnx/defs/shape_inference.h [hunk 2]
```diff
@@ -172,94 +176,108 @@ inline void updateOutputElemType(
...
-inline void updateOutputShape(
+inline Status updateOutputShape(
     InferenceContext& ctx,
     size_t outputIndex,
     const TensorShapeProto& shape) {
   auto* output_shape = getOutputShape(ctx, outputIndex);
-  if (output_shape != nullptr) {
-    *output_shape = shape;
+  if (output_shape == nullptr) {
+	return Status(INFERENCE, INVALID_PROTOBUF, MakeString("Output: ", outputIndex, " should be tensor-type or no-set."));
```
Comment (gramalingam) [inline onnx/defs/shape_inference.h]: I think this won't work. We need to allow optional outputs (which may be missing, and that is valid). Unfortunately, we can distinguish between the two cases only inside "getOutputShape".

**Comment by anderspapitto** [comment]
@gramalingam the implementation we provide (in implementation.h) has the behavior of

- getOutputType(n) is always a valid, fresh pointer (as long as n < numOutputs)

I think we shouldn't have to really deal with optional inputs/outputs - by the time shape inference runs, there already is some complete graph, and each node we see has a fixed number of inputs/outputs. The shape inference function should simply infer shapes for as many outputs as it can.

File: onnx/defs/experiments/defs.cc (hunk 1 above)
Comment (linkerzhang) [reply onnx/defs/experiments/defs.cc]: I may remove INVALID_ARGUMENT if it's confusing. It's not used in this PR.

File: onnx/common/status.cc [hunk 3]
```diff
@@ -40,6 +40,8 @@ std::string Status::ToString() const {
     result += "[CheckerError]";
   } else if (StatusCategory::OPTIMIZER == state_->category) {
     result += "[OptimizerError]";
+  } else if (StatusCategory::INFERENCE == state_->category) {
+	result += "[InferenceError]";
```
Comment (anderspapitto) [inline onnx/common/status.cc]: I would suggest "ShapeInferenceError" in a user-facing message, since Inference is an overloaded term

**Comment by gramalingam** [comment]
@anderspapitto : ok (but this just hides optional non-trailing outputs; for optional trailing outputs, the inference code still needs to check "numOutputs" to figure out if that is present or not)

File: onnx/defs/shape_inference.h (hunk 2 above)
Comment (gramalingam) [reply onnx/defs/shape_inference.h]: Ok, please ignore my comment above, as per discussion elsewhere

Comment (gramalingam) [reply onnx/defs/shape_inference.h]: But, essentially, this condition will never arise as per our discussion.

**Comment by gramalingam** [comment]
@anderspapitto : is there a potential problem in implementation.h? if an optional output indicated by an empty string name has an inferred type, then it gets saved as the type of the empty string, and may be used as the type of another optional input/output elsewhere ... at least implementation.h needs to handle input/outputs with an empty string name, right?

File: onnx/common/status.cc (hunk 3 above)
Comment (linkerzhang) [reply onnx/common/status.cc]: But type inference also covered in the function :). So, TypeShapeInference? if you want.

File: onnx/defs/tensor/defs.cc [hunk 4]
```diff
@@ -251,27 +254,27 @@ Otherwise, the tensor is split to equal sized parts.
...
     .TypeAndShapeInferenceFunction([](InferenceContext& ctx) {
       for (int i = 0; i < static_cast<int>(ctx.getNumOutputs()); ++i) {
-        propagateElemTypeFromInputToOutput(ctx, 0, i);
+        ONNX_RETURN_IF_ERROR(propagateElemTypeFromInputToOutput(ctx, 0, i));
       }
 
-      if (!hasNInputShapes(ctx, 1)) {
-        return;
-      }
+	  if (hasNInputShapes(ctx, 1)) {
```
Comment (gramalingam) [inline onnx/defs/tensor/defs.cc]: Missing !

Comment (linkerzhang) [reply onnx/defs/tensor/defs.cc]: Thank you so much for the catch and fix!

**Comment by bddppq** [comment]
@linkerzhang The CI reported there are core dump in the tests, probably some nullptr dereferencing?

**Comment by anderspapitto** [comment]
@gramalingam yes, there might be an issue like that. Do we currently have any operators with optional, non-trailing outputs? We may actually want to also just come up with a cleaner representation for optional inputs/outputs than "empty string name".

**Comment by dzhulgakov** [comment]
Hey folks, maybe it was discussed before - but why status codes instead of exceptions? It's much easier to propagate errors that way and we already do use exceptions in onnx codebase

**Comment by linkerzhang** [comment]
@dzhulgakov I thought we discussed this before that we agreed to have "Status" introduced into ONNX for error handling. Not all upper callers would like to follow using exception for error handling. For shape inference specifically, it's not an exception if it fails. Runtimes may still be able to handle the case without shape inference.

**Comment by bddppq** [comment]
@linkerzhang If shape inference throws exception, can your code try && catch? I'm really not trying to say exception is better than status code in general, it will be a endless debate. My point is onnx started with using exceptions, there is no good reason to do such an api breaking change, and we definitely need to have a consistent style of error checking.

**Pull request closed without merging**